blockstore/blockstore.go

```diff
@@ -310,6 +317,227 @@ func (bs *blockstore) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
 	return output, nil
 }
 
+// GetManyOption is a getManyBlockStore option implementation
+type GetManyOption struct {
+	f func(bs *getManyBlockStore)
+}
+
+// NewGetManyBlockstore returns a default GetManyBlockstore implementation
+// using the provided datastore.TxnDatastore backend.
+func NewGetManyBlockstore(d ds.TxnDatastore, opts ...GetManyOption) GetManyBlockstore {
+	bs := &getManyBlockStore{
+		datastore: d,
+	}
+
+	for _, o := range opts {
+		o.f(bs)
+	}
+
+	if !bs.noPrefix {
+		bs.datastore = dsns.WrapTxnDatastore(bs.datastore, BlockPrefix)
+	}
+	return bs
+}
+
+type getManyBlockStore struct {
+	datastore ds.TxnDatastore
+
+	rehash       atomic.Bool
+	writeThrough bool
+	noPrefix     bool
+}
+
+func (bs *getManyBlockStore) HashOnRead(enabled bool) {
+	bs.rehash.Store(enabled)
+}
+
+func (bs *getManyBlockStore) Get(ctx context.Context, k cid.Cid) (blocks.Block, error) {
```

Comment:
I am surprised this function needs to be duplicated, I think you could use embeding here.

Reply:
You are 100% correct- will adjust.

Reply:
This is possible but would require passing in a `ds.Batching` to the constructor function in addition to the`ds.TxnDatastore` as `ds.TxnDatastore` can't be used to construct a regular `Blockstore` for embedding because it doesn't satisfy the `ds.Batching` interface, even though the `Put` and `Get` methods we would like to fall through to don't use the portions of that interface that it doesn't support...

Reply:
Another option would be to break the normal `Blockstore` up into the components that need `ds.Batching` and the ones that just need `ds.Datastore`, `ds.TxDatastore` satisfies `ds.Datastore` and we only need `ds.Datastore` to fulfill the `Get` and `Put` methods on either blockstore.

Reply:
Let me know what you think the best/cleanest approach is here

Reply:
I would make it take the union of batching and txn datastores so it has access to all that is needed.
You could also implement a stub, a batch is inferior to a transaction while providing the same features:
> Batching datastores support deferred, grouped updates to the database. `Batch`es do NOT have transactional semantics: updates to the underlying datastore are not guaranteed to occur in the same iota of time. Similarly, batched updates will not be flushed to the underlying datastore until `Commit` has been called. `Txn`s from a `TxnDatastore` have all the capabilities of a `Batch`, but the reverse is NOT true. 

https://pkg.go.dev/github.com/ipfs/go-datastore#Batching

So if you accept a TxnDatastore you can use a simple stub that implements batches using transactions, this would make thee code work in the only place it's used right now.